### PR TITLE
support network ipam for compose

### PR DESCRIPTION
**Pull request by Junnplus**
Signed-off-by: ye.sijun <junnplus@gmail.com>

Fixes: https://github.com/containerd/nerdctl/issues/432
Thank https://github.com/containerd/nerdctl/pull/436

### CODE REVIEW

File: cmd/nerdctl/compose_up_linux_test.go [hunk 1]
```diff
@@ -298,3 +298,31 @@ services:
 
 	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps").AssertOutContains(fmt.Sprintf("%s_test_2", projectName))
 }
+
+func TestIPAMConfig(t *testing.T) {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/compose_up_linux_test.go:302]:
```suggestion
func TestComposeUpIPAMConfig(t *testing.T) {
```

File: pkg/composer/up_network.go [hunk 2]
```diff
@@ -51,8 +51,13 @@ func (c *Composer) upNetwork(ctx context.Context, shortName string) error {
 		createArgs := []string{
 			fmt.Sprintf("--label=%s=%s", labels.ComposeProject, c.Options.Project),
 			fmt.Sprintf("--label=%s=%s", labels.ComposeNetwork, shortName),
-			fullName,
 		}
+
+		for _, ipamConfig := range net.Ipam.Config {
```
Comment (AkihiroSuda) [inline pkg/composer/up_network.go:56]:
```go
	if unknown := reflectutil.UnknownNonEmptyFields(&ipamConfig, "Subnet"); len(unknown) > 0 {
		logrus.Warnf("Ignoring: network %s: ipam.config[%d]: %+v", shortName, i, unknown)
	}
```

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: cmd/nerdctl/compose_up_linux_test.go [hunk 3]
```diff
@@ -298,3 +298,31 @@ services:
...
+  foo:
+    image: alpine
+    command: "sleep infinity"
+
+networks:
+  default:
+    ipam:
+      config:
+        - subnet: 10.1.100.0/24
+`
```
Comment (fahedouch) [inline cmd/nerdctl/compose_up_linux_test.go:316]:
```
var dockerComposeYAML = fmt.Sprintf(`
version: '3.1'

services:
  foo:
    image: %s
    command: "sleep infinity"
networks:
  default:
    ipam:
      config:
        - subnet: 10.1.100.0/24
`, testutil.AlpineImage)
```

**Comment by fahedouch** [comment]
LGTM on green CI

**Comment by AkihiroSuda** [comment]
```
== RUN   TestComposeIPAMConfig
    compose_up_linux_test.go:322: projectName="nerdctl-compose-test2542743296"
    compose_up_linux_test.go:324: assertion failed: expression is false: res.ExitCode == exitCode: time="2021-12-27T09:47:25Z" level=info msg="Creating network nerdctl-compose-test2542743296_default"
        panic: reflect: call of reflect.Value.NumField on ptr Value
        goroutine 1 [running]:
        reflect.flag.mustBe(...)
        	/usr/local/go/src/reflect/value.go:223
        reflect.Value.NumField({0x10670a0, 0xc00059c880, 0x1290cf6})
        	/usr/local/go/src/reflect/value.go:1710 +0x8d
        github.com/containerd/nerdctl/pkg/reflectutil.UnknownNonEmptyFields({0x1062b60, 0xc00059c880}, {0xc0006df520, 0x2, 0x2})
        	/go/src/github.com/containerd/nerdctl/pkg/reflectutil/reflectutil.go:40 +0x23d
        github.com/containerd/nerdctl/pkg/composer.(*Composer).upNetwork(0xc0001aa2a0, {0x1599360, 0xc000114000}, {0xc0005fda10, 0x7})
        	/go/src/github.com/containerd/nerdctl/pkg/composer/up_network.go:57 +0x65b
        github.com/containerd/nerdctl/pkg/composer.(*Composer).Up(0xc0001aa2a0, {0x1599360, 0xc000114000}, {0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
        	/go/src/github.com/containerd/nerdctl/pkg/composer/up.go:44 +0xea
        main.composeUpAction(0xc000323400, {0xc000518fa0, 0x0, 0x1})
        	/go/src/github.com/containerd/nerdctl/cmd/nerdctl/compose_up.go:117 +0x53c
        github.com/spf13/cobra.(*Command).execute(0xc000323400, {0xc00003a0d0, 0x1, 0x1})
        	/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:856 +0x60e
        github.com/spf13/cobra.(*Command).ExecuteC(0xc0001b7680)
        	/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bc
        github.com/spf13/cobra.(*Command).Execute(...)
        	/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:902
        main.xmain()
        	/go/src/github.com/containerd/nerdctl/cmd/nerdctl/main.go:84 +0x9f
        main.main()
        	/go/src/github.com/containerd/nerdctl/cmd/nerdctl/main.go:67 +0x19
--- FAIL: TestComposeIPAMConfig (0.02s)
```

https://cirrus-ci.com/task/4883230380785664?logs=rootful_test#L215

**Comment by AkihiroSuda** [comment]
```
--- PASS: TestComposeUpWithScale (0.53s)
=== RUN   TestComposeIPAMConfig
    compose_up_linux_test.go:324: projectName="nerdctl-compose-test2749460603"
    compose_up_linux_test.go:326: assertion failed: expression is false: res.ExitCode == exitCode: time="2021-12-27T10:48:36Z" level=info msg="Creating network nerdctl-compose-test2749460603_default"
        time="2021-12-27T10:48:37Z" level=info msg="Ensuring image ghcr.io/stargz-containers/alpine:3.13-org"
        time="2021-12-27T10:48:37Z" level=info msg="Creating container nerdctl-compose-test2749460603_foo_1"
        time="2021-12-27T10:48:37Z" level=fatal msg="failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: time=\"2021-12-27T10:48:37Z\" level=fatal msg=\"failed to call cni.Setup: plugin type=\\\"bridge\\\" failed (add): failed to set bridge addr: \\\"nerdctl1\\\" already has an IP address different from 10.1.100.1/24\"\nFailed to write to log, write /var/lib/nerdctl/1935db59/containers/nerdctl-test/0c6b3bc6622c1ec0cead5d768ad577f8d0347c0b0c74501c98aef03d85290411/oci-hook.createRuntime.log: file already closed: unknown"
        time="2021-12-27T10:48:37Z" level=fatal msg="error while creating container nerdctl-compose-test2749460603_foo_1: exit status 1"
--- FAIL: TestComposeIPAMConfig (0.24s)
```

https://cirrus-ci.com/task/4506924338118656?logs=rootful_test#L214

**Comment by Junnplus** [comment]
@AkihiroSuda It seems that `network rm` is not completely

**Comment by fahedouch** [comment]
I think that problem come from the composeScale test https://github.com/containerd/nerdctl/blob/ed3d553a9bd6818b0917f2e040e43df5973ff8c2/cmd/nerdctl/compose_up_linux_test.go#L297
Scaled replicas are not removed with defer because they are not mentioned on docker-compose.yml but the svc itself os removed. This orphan cni  config should make this error. I will try to test this tomorrow and make a fix to remove orphan container in the Compose svc

**Comment by AkihiroSuda** [comment]
Please rebase

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**